### PR TITLE
The game begins

### PR DESCRIPTION
**Pull request by peara**

### CODE REVIEW

**Comment by pinkiebell** [comment]
@peara 
Discovery 🌮 🎉 
Now this looks like a problem. Do you have a potential solution in mind ? 🙂

**Comment by peara** [comment]
@pinkiebell 
I think we can solve this by require an additional `submitProof`.
This one need to prove the existence of the initial state.

When at a leaf:
- `submitPreviousHash(merkleProofs)`: position should be right before the contested leaf. Let's call this `prevLeaf`.
- `submitProofPrev()`: prove `prevLeaf`.
- `submitProof()`: prove `leaf`

But this could be a burden for solver, as these transactions are not cheap. So if this works, maybe we should also raise dispute bond of challenger.

**Comment by pinkiebell** [comment]
I had almost the same thing in mind.
The second solution I came up with is (pseudocode):
- add `bytes32 witness` to the dispute struct
- whenever we go right: we update `dispute.witness = computationPath.left`
- if we get to the point in `updateRound` where `dispute.treeDepth == 0 && dispute.challenger.left != dispute.solver.left`
   we check `dispute.state & START_OF_EXECUTION) != 0`, if true we don't care and `submitProof` will handle this case
else we need:
     `witnesses.length >= 2 && witnesses.length % 2 == 0` where `witnesses` is a `bytes32[]`
      needs to be supplied by the sender
- we hash the witnesses and verify that it matches the `witness`-hash
- from these proofs we can verify which is the correct `computationPath.left`

**Comment by johannbarbie** [comment]
if the tree is too complicated, why not simply copy the way truebit does it? 
https://github.com/TrueBitFoundation/truebit-os/blob/master/contracts/dispute/Interactive.sol

**Comment by peara** [comment]
It was designed to save gas. If this is not solvable, we will use truebit's way.

**Comment by peara** [comment]
@pinkiebell 
I have updated some test cases. It failed because of large witnesses.

File: contracts/Verifier.sol [hunk 1]
```diff
@@ -338,25 +337,40 @@ contract Verifier is Ownable, HydratedRuntime {
...
+            dispute.state |= SOLVER_VERIFIED;
+            return;
+        }
+
         if (dispute.treeDepth == 1 && dispute.challenger.left != dispute.solver.left) {
             require(dispute.witness != bytes32(0));
-            // we may only require 6 or 2 witnesses
-            // this could be improved
-            require(witnesses.length == 6 || witnesses.length == 2);
+            // TODO: we could keep track of the last depth taken right,
```
Comment (peara) [inline contracts/Verifier.sol]: I suppose we could do this by:
- when go right, witness = left
- when go left, witness = witness.right, need to respond with (witness.left, witness.right) to verify.

Comment (pinkiebell) [reply contracts/Verifier.sol]: @peara True 😺. Updated 😏

**Comment by peara** [comment]
@pinkiebell 
Now this looks great :smile: 
I haven't thought up anything else.
#107 is also solved.

**Pull request merged**